File: android/src/main/java/ch/nadlo/oss/capacitor/sip_phone/SipPhoneControl.java
package ch.nadlo.oss.capacitor.sip_phone;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.collection.ArraySet;

import org.linphone.core.Account;
import org.linphone.core.AccountParams;
import org.linphone.core.Address;
import org.linphone.core.AudioDevice;
import org.linphone.core.Call;
import org.linphone.core.Core;
import org.linphone.core.CoreListener;
import org.linphone.core.CoreListenerStub;
import org.linphone.core.Factory;
import org.linphone.core.RegistrationState;


public class SipPhoneControl {

    private static final String TAG = "SipPhoneControl";
    private final Context context;
    private final ArraySet<SipPhoneEventListener> listeners = new ArraySet<>();

    public boolean isLoggedIn = false;
    public boolean isCallRunning = false;
    public boolean isCallIncoming = false;
    public boolean isCallOutgoing = false;
    private final CoreListener coreListener = new CoreListenerStub() {
        @Override
        public void onAccountRegistrationStateChanged(@NonNull Core core, @NonNull Account account, RegistrationState state, @NonNull String message) {
            Log.i(TAG, "onAccountRegistrationStateChanged: state=" + state.toString() + " (" + message + ")");
            isLoggedIn = state == RegistrationState.Ok;

            // DEBUG
            Log.v(TAG, "onAccountRegistrationStateChanged: isPushNotificationAvailable " + core.isPushNotificationAvailable());

            for (var listener : listeners) {
                listener.onAccountStateChanged();
            }
        }

        @Override
        public void onAudioDeviceChanged(@NonNull Core core, @NonNull AudioDevice audioDevice) {
            // This callback will be triggered when a successful audio device has been changed
            Log.i(TAG, "onAudioDeviceChanged: " + audioDevice);
        }

        @Override
        public void onAudioDevicesListUpdated(@NonNull Core core) {
            // This callback will be triggered when the available devices list has changed,
            // for example after a bluetooth headset has been connected/disconnected.
            Log.i(TAG, "onAudioDevicesListUpdated");
        }

        @Override
        public void onCallStateChanged(@NonNull Core core, Call call, Call.State state, @NonNull String message) {
            Log.i(TAG, "onCallStateChanged: " + call.getRemoteContact() + " state=" + state.toString() + " (" + message + ")");
            isCallRunning = false;
            isCallOutgoing = false;
            isCallIncoming = false;
            switch (state) {
                case Connected, StreamsRunning -> isCallRunning = true;
                case PushIncomingReceived, IncomingReceived -> isCallIncoming = true;
                case OutgoingInit, OutgoingProgress, OutgoingRinging -> isCallOutgoing = true;
            }
            for (var listener : listeners) {
                listener.onCallStateChanged();
            }
        }
    };
    private SipLoginOptions options;
    private Core core;

    public SipPhoneControl(Context context) {
        this.context = context;
    }

    public void login(SipLoginOptions options) {
        this.options = options;
        initCore();
        setSipLoginOptions();
        core.start();

        if (!core.isPushNotificationAvailable()) {
            // TODO ???
            Log.w(TAG, "linPhoneLogin: isPushNotificationAvailable IS FALSE ");
        }

        Log.i(TAG, "linPhoneLogin done.");
    }

    public void start() {
        if (core == null) {
            initCore();
        }
        core.start();
    }

    public void logout() {
//        core.stop(); // ???
        unregister();
        delete();
    }

    private void unregister() {
        // Here we will disable the registration of our Account
        Account account = core.getDefaultAccount();
        if (account == null) {
            Log.w(TAG, "unregister failed no default account");
            return;
        }

        AccountParams params = account.getParams();
        // Returned params object is const, so to make changes we first need to clone it
        AccountParams clonedParams = params.clone();

        // Now let's make our changes
        clonedParams.setRegisterEnabled(false);

        // And apply them
        account.setParams(clonedParams);
    }

    private void delete() {
        // To completely remove an Account
        Account account = core.getDefaultAccount();
        if (account == null) {
            Log.w(TAG, "delete - no default account");
            return;
        } else {
            core.removeAccount(account);
        }

        // To remove all accounts use
        core.clearAccounts();
        // Same for auth info
        core.clearAllAuthInfo();
    }

    private void initCore() {
        Factory.instance().setDebugMode(true, "LinPhoneCapPlugin");
        core = Factory.instance().createCore("LinPhone-core-config", null, context);
        core.addListener(coreListener);

        // For push notifications to work, you have to copy your google-services.json in the app/ folder
        // And you must declare our FirebaseMessaging service in the Manifest
        // You also have to make some changes in your build.gradle files, see the ones in this project
        // Make sure the core is configured to use push notification token from firebase
        core.setPushNotificationEnabled(true);
    }

    private void setSipLoginOptions() {
        var authInfo = Factory.instance().createAuthInfo(options.username, null, options.password, null, null, options.domain, null);

        var params = core.createAccountParams();
        var identity = Factory.instance().createAddress("sip:" + options.username + "@" + options.domain);
        params.setIdentityAddress(identity);

        var address = Factory.instance().createAddress("sip:" + options.domain);
        address.setTransport(options.transportType);

        params.setServerAddress(address);
        params.setRegisterEnabled(true);
        // Ensure push notification is enabled for this account
        params.setPushNotificationAllowed(true);

        core.addAuthInfo(authInfo);
        var account = core.createAccount(params);
        core.addAccount(account);
        core.setDefaultAccount(account);
    }


    private void toggleSpeaker() {
        // Get the currently used audio device
        var currentCall = core.getCurrentCall();
        assert currentCall != null;
        var audioDevice = currentCall.getOutputAudioDevice();
        assert audioDevice != null;
        var speakerEnabled = audioDevice.getType() == AudioDevice.Type.Speaker;

        // We can get a list of all available audio devices using
        // Note that on tablets for example, there may be no Earpiece device

//        for (audioDevice in core.getAudioDevices()) {
//            if (speakerEnabled && audioDevice.type == AudioDevice.Type.Earpiece) {
//                core.currentCall?.outputAudioDevice = audioDevice
//                return
//            } else if (!speakerEnabled && audioDevice.type == AudioDevice.Type.Speaker) {
//                core.currentCall?.outputAudioDevice = audioDevice
//                return
//            }/* If we wanted to route the audio to a bluetooth headset
//            else if (audioDevice.type == AudioDevice.Type.Bluetooth) {
//                core.currentCall?.outputAudioDevice = audioDevice
//            }*/
//        }
    }

    public void acceptCurrentCall() {
        var call = core.getCurrentCall();
        if (call == null) {
            Log.e(TAG, "acceptCurrentCall: no call");
            throw new RuntimeException("No call");
        }
        Log.i(TAG, "acceptCurrentCall: call state=" + call.getState());
        try {
            call.accept();
        } catch (Exception err) {
            Log.e(TAG, "acceptCurrentCall: " + err.getMessage(), err);
            throw err;
        }
    }

    public void terminateCurrentCall() {
        var call = core.getCurrentCall();
        if (call == null) {
            Log.e(TAG, "terminateCurrentCall: no call");
            throw new RuntimeException("No call");
        }
        Log.i(TAG, "terminateCurrentCall: call state=" + call.getState());
        try {
            call.terminate();
        } catch (Exception err) {
            Log.e(TAG, "terminateCurrentCall: " + err.getMessage(), err);
            throw err;
        }
    }

    public void addListener(SipPhoneEventListener listener) {
        this.listeners.add(listener);
    }

    public SipAddress getCurrentCallRemoteAddress() {
        var remoteAddress = core.getCurrentCallRemoteAddress();
        if (remoteAddress != null) {
            Log.v(TAG, "getCurrentCallRemoteAddress: " + remoteAddress.asString());
            return new SipAddress(remoteAddress);
        }
        Log.v(TAG, "getCurrentCallRemoteAddress: no active call");
        return null;
    }

    static public class SipAddress {
        public String uri;
        public String displayName;
        public String username;

        SipAddress(Address remoteAddress) {
            displayName = remoteAddress.getDisplayName();
            uri = remoteAddress.asStringUriOnly();
            username = remoteAddress.getUsername();
        }
    }
}

File: ios/Plugin/SipPhoneControl.swift
import Foundation
import linphonesw

class SipPhoneControl: ObservableObject {
    var mCore: Core!
    var mAccount: Account?
    var mCoreDelegate : CoreDelegate!

    @Published var username : String = ""
    @Published var passwd : String = ""
    @Published var domain : String = ""
    @Published var loggedIn: Bool = false
    @Published var transportType : String = "TCP"

    @Published var currentAccount : String = ""

    // Outgoing call related variables
    @Published var callMsg : String = ""
    @Published var isCallRunning : Bool = false
    @Published var isVideoEnabled : Bool = false
    @Published var canChangeCamera : Bool = false
    @Published var remoteAddress : String = ""

    // Callkit related variables
    @Published var isCallIncoming: Bool = false
    @Published var isCallOutgoing: Bool = false
    @Published var incomingCallName = ""
    @Published var mCall : Call?
    @Published var mProviderDelegate : SipPhoneCallKitProviderDelegate!
    @Published var mCallAlreadyStopped : Bool = false

    // registration tokens
    @Published var tokenVOIP : String = ""
    @Published var tokenPUSH : String = ""

    // event handlers
    @Published var registrationStateListener: (() -> Void)?;
    @Published var callStateListener: (() -> Void)?;

    init()
    {
        LoggingService.Instance.logLevel = LogLevel.Debug

        let factory = Factory.Instance
        // IMPORTANT : In this tutorial, we require the use of a core configuration file.
        // This way, once the registration is done, and until it is cleared, it will return to the LoggedIn state on launch.
        // This allows us to have a functional call when the app was closed and is started by a VOIP push notification (incoming call
        // We also need to enable "Push Notitifications" and "Background Mode - Voice Over IP"
        let configDir = factory.getConfigDir(context: nil)
        try? mCore = factory.createCore(configPath: "\(configDir)/SipConfig", factoryConfigPath: "", systemContext: nil)

//        mCore.limeX3DhEnabled = false;

        mProviderDelegate = SipPhoneCallKitProviderDelegate(context: self)

        // enabling push notifications management in the core

        /// Special function to enable the callkit.
        mCore.callkitEnabled = true
        /// Enable or disable push notifications on Android & iOS.
        /// If enabled, it will try to get the push token add configure each account with
        /// push_notification_allowed set to true with push parameters. IOS: will also
        /// instanciate a PushRegistry, so make sure that your app does not instanciate one
        /// too or there will be a conflict.
        mCore.pushNotificationEnabled = true

        // call this when pushes are handled manually
//        mCore.processPushNotification(callId: "")

        mCore.pushNotificationConfig?.teamId = "7485TYL3QC";

        NSLog("[SipPhoneControl] init teamId=" + (mCore.pushNotificationConfig?.teamId ?? "nil"))
        NSLog("[SipPhoneControl] init bundleIdentifier=" + ( mCore.pushNotificationConfig?.bundleIdentifier ?? "nil"))
        NSLog("[SipPhoneControl] init remoteToken=" + (mCore.pushNotificationConfig?.remoteToken ?? "nil"))
        NSLog("[SipPhoneControl] init voipToken=" + (mCore.pushNotificationConfig?.voipToken ?? "nil"))

        // Here we enable the video capture & display at Core level
        // It doesn't mean calls will be made with video automatically,
        // But it allows to use it later
        mCore.videoCaptureEnabled = false
        mCore.videoDisplayEnabled = true

        // When enabling the video, the remote will either automatically answer the update request
        // or it will ask it's user depending on it's policy.
        // Here we have configured the policy to always automatically accept video requests
        mCore.videoActivationPolicy!.automaticallyAccept = true

        // If the following property is enabled, it will automatically configure created call params with video enabled
        mCore.videoActivationPolicy!.automaticallyInitiate = true

        // ???
        // mCore.videoEnabled = true;

        NSLog("[SipPhoneControl] init before start")

        try? mCore.start()

        NSLog("[SipPhoneControl] init after start")

        mCoreDelegate = CoreDelegateStub( onCallStateChanged: { (core: Core, call: Call, state: Call.State, message: String) in
            // This function will be called each time a call state changes,
            // which includes new incoming/outgoing calls
            self.callMsg = message

            NSLog("[SipPhoneControl] onCallStateChanged callMsg \(message)")
            NSLog("[SipPhoneControl] onCallStateChanged state \(state)")
            NSLog("[SipPhoneControl] onCallStateChanged remoteContact " + call.remoteContact)


            if (state == .PushIncomingReceived){
                // We're being called by someone (and app is in background)
                self.mCall = call
                self.isCallIncoming = true

                self.incomingCallName = call.remoteAddress?.displayName ?? "Unknown"
//                self.remoteAddress = call.remoteAddress!.asStringUriOnly()

                self.mProviderDelegate.incomingCall()
            } else if (state == .IncomingReceived) {
                // If app is in foreground, it's likely that we will receive the SIP invite before the Push notification
                if (!self.isCallIncoming) {
                    self.mCall = call
                    self.isCallIncoming = true

                    self.incomingCallName = call.remoteAddress?.displayName ?? "Unknown"

                    self.mProviderDelegate.incomingCall()
                }
                self.remoteAddress = call.remoteAddress!.asStringUriOnly()
            } else if (state == .OutgoingInit) {
                // When the 200 OK has been received
                self.isCallIncoming = false
                self.isCallOutgoing = true

                // repot outgoing call
                self.mProviderDelegate.outgoingCallStarted()
            } else if (state == .OutgoingRinging) {
                // This state will be reached upon reception of the 180 RINGING
            } else if (state == .Connected) {
                // When the 200 OK has been received
                self.isCallIncoming = false
                self.isCallOutgoing = false
                self.isCallRunning = true

                self.mProviderDelegate.stopCall(_callHandledInApp: true)

            } else if (state == .StreamsRunning) {
                // This state indicates the call is active.
                // You may reach this state multiple times, for example after a pause/resume
                // or after the ICE negotiation completes

                // Only enable toggle camera button if there is more than 1 camera
                // We check if core.videoDevicesList.size > 2 because of the fake camera with static image created by our SDK (see below)
                self.canChangeCamera = core.videoDevicesList.count > 2

                // report like connected
                if (self.isCallOutgoing) {
                    self.mProviderDelegate.outgoingCallConnected();
                }
            } else if (state == .Paused) {
                // When you put a call in pause, it will became Paused
                self.canChangeCamera = false
            } else if (state == .PausedByRemote) {
                // When the remote end of the call pauses it, it will be PausedByRemote
            } else if (state == .Updating) {
                // When we request a call update, for example when toggling video
            } else if (state == .UpdatedByRemote) {
                // When the remote requests a call update
            } else if (state == .Released || state == .End || state == .Error) {
                // Call has been terminated by any side

                // Report to CallKit that the call is over, if the terminate action was initiated by other end of the call
                if (self.isCallRunning) {
                    self.mProviderDelegate.stopCall(_callHandledInApp: false)
                }

                self.remoteAddress = ""

                self.isCallRunning = false
                self.isCallOutgoing = false
                self.isCallIncoming = false
            }

            // notify changed call state
            if let callback = self.callStateListener {
                callback()
            }
        }, onAccountRegistrationStateChanged: { (core: Core, account: Account, state: RegistrationState, message: String) in
            NSLog("[SipPhoneControl] New registration state is \(state) for user id \( String(describing: account.params?.identityAddress?.asString()))\n")


            if (state == .Ok) {
                self.loggedIn = true
                if(self.mAccount != nil) {
                    self.mAccount = account
                }
                self.currentAccount = account.contactAddress?.username ?? ""
            } else if (state == .Cleared) {
                self.loggedIn = false
            }

            if let callback = self.registrationStateListener {
                callback()
            } else {
                NSLog("[SipPhoneControl] onAccountRegistrationStateChanged - callback not found")
                // TODO schedule later
            }
        })

        mCore.addDelegate(delegate: mCoreDelegate)
    }

    func login() throws {
        do {
            if (self.loggedIn) {
                NSLog("[SipPhoneControl] Login - already logged in")
                if let callback = self.registrationStateListener {
                    callback()
                } else {
                    NSLog("[SipPhoneControl] Login - callback not found")
                }
                return
            }

            var transport : TransportType
            if (transportType == "TLS") { transport = TransportType.Tls }
            else if (transportType == "TCP") { transport = TransportType.Tcp }
            else  { transport = TransportType.Udp }

            transport = TransportType.Tcp

            NSLog("[SipPhoneControl] Login \(transport):\\\\\(username)@\(domain)")

            let authInfo = try Factory.Instance.createAuthInfo(username: username, userid: "", passwd: passwd, ha1: "", realm: "", domain: domain)
            let accountParams = try mCore.createAccountParams()
            let identity = try Factory.Instance.createAddress(addr: String("sip:" + username + "@" + domain))
            try! accountParams.setIdentityaddress(newValue: identity)
            let address = try Factory.Instance.createAddress(addr: String("sip:" + domain))
            try address.setTransport(newValue: transport)
            try accountParams.setServeraddress(newValue: address)
            accountParams.registerEnabled = true

            NSLog("[SipPhoneControl] Login teamId=" + (accountParams.pushNotificationConfig?.teamId ?? "nil"))
            NSLog("[SipPhoneControl] Login bundleIdentifier=" + ( accountParams.pushNotificationConfig?.bundleIdentifier ?? "nil"))
            NSLog("[SipPhoneControl] Login remoteToken=" + (accountParams.pushNotificationConfig?.remoteToken ?? "nil"))
            NSLog("[SipPhoneControl] Login voipToken=" + (accountParams.pushNotificationConfig?.voipToken ?? "nil"))
            NSLog("[SipPhoneControl] Login pushNotificationAllowed=" + (accountParams.pushNotificationAllowed ? "Y" : "N" ))
            NSLog("[SipPhoneControl] Login remotePushNotificationAllowed=" + (accountParams.remotePushNotificationAllowed  ? "Y" : "N"))

            tokenPUSH = accountParams.pushNotificationConfig?.remoteToken ?? ""
            tokenVOIP = accountParams.pushNotificationConfig?.voipToken ?? ""

            NSLog("[SipPhoneControl] Login update account params...")
            accountParams.pushNotificationAllowed = false;
            accountParams.remotePushNotificationAllowed = false;

            NSLog("[SipPhoneControl] Login pushNotificationAllowed=" + (accountParams.pushNotificationAllowed ? "Y" : "N" ))
            NSLog("[SipPhoneControl] Login remotePushNotificationAllowed=" + (accountParams.remotePushNotificationAllowed  ? "Y" : "N"))

            // Enable push notifications on this account
            // https://wiki.linphone.org/xwiki/wiki/public/view/Lib/Features/Push%20notifications/IOS%20push%20notification%20management
//            accountParams.pushNotificationAllowed = false /// Indicates whether to add to the contact parameters the push notification information.  For IOS, it indicates for VOIP push notification.
//            accountParams.remotePushNotificationAllowed = false /// Indicates whether to add to the contact parameters the push notification information.
            // NOTE: enable only one to support FreePBX calls (fails due to & mark)


            // We're in a sandbox application, so we must set the provider to "apns.dev" since it will be "apns" by default, which is used only for production apps
//            accountParams.pushNotificationConfig?.provider = "apns.dev";


            NSLog("[SipPhoneControl] Login create account...")

            mAccount = try mCore.createAccount(params: accountParams)
            mCore.addAuthInfo(info: authInfo)
            try mCore.addAccount(account: mAccount!)
            mCore.defaultAccount = mAccount

        } catch {
            NSLog("[SipPhoneControl] Login error \(error.localizedDescription)")
            throw error
        }
    }

    func unregister() throws
    {
        if let account = mCore.defaultAccount {
            NSLog("[SipPhoneControl] unregister: " + (account.contactAddress?.username ?? "<contactAddress>"))
            let params = account.params
            let clonedParams = params?.clone()
            clonedParams?.registerEnabled = false
            account.params = clonedParams
        } else {
            NSLog("[SipPhoneControl] unregister - no default account")
        }
        loggedIn = false
    }

    func delete()  throws {
        if let account = mCore.defaultAccount {
            NSLog("[SipPhoneControl] delete: " + (account.contactAddress?.username ?? "<contactAddress>"))
            mCore.removeAccount(account: account)
        }
        NSLog("[SipPhoneControl] delete - clearAccounts")
        mCore.clearAccounts()
        mCore.clearAllAuthInfo()
        loggedIn = false
    }


    func outgoingCall() throws {
        do {
            // As for everything we need to get the SIP URI of the remote and convert it to an Address
            let remoteAddress = try Factory.Instance.createAddress(addr: remoteAddress)

            // We also need a CallParams object
            // Create call params expects a Call object for incoming calls, but for outgoing we must use null safely
            let params = try mCore.createCallParams(call: nil)

            // We can now configure it
            // Here we ask for no encryption but we could ask for ZRTP/SRTP/DTLS
            params.mediaEncryption = MediaEncryption.None
            // If we wanted to start the call with video directly
            //params.videoEnabled = true

            // Finally we start the call
            let _ = mCore.inviteAddressWithParams(addr: remoteAddress, params: params)
            // Call process can be followed in onCallStateChanged callback from core listener
        } catch {
            throw error
        }

    }

    func terminateCall() {
        if (self.isCallRunning) {
            do {
                if (mCore.callsNb == 0) { return }

                // If the call state isn't paused, we can get it using core.currentCall
                let coreCall = (mCore.currentCall != nil) ? mCore.currentCall : mCore.calls[0]

                // Terminating a call is quite simple
                if let call = coreCall {
                    try call.terminate()
                }
            } catch { NSLog(error.localizedDescription) }
        }
    }

    func registerRemotePush(deviceToken: Data) {
        var stringifiedToken = deviceToken.map{String(format: "%02X", $0)}.joined()
        stringifiedToken.append(String(":remote"))
        mCore.didRegisterForRemotePushWithStringifiedToken(deviceTokenStr: stringifiedToken)
        NSLog("[SipPhoneControl] push-token registered \(stringifiedToken)")
    }
}
